ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/LedgerTables.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Test.Consensus.Cardano.LedgerTables (tests) where
+
+import           Data.Map.Diff.Strict (Values (..))
+import qualified Data.Map.Strict as Map
+import           Data.SOP.Strict
+
+import           Ouroboros.Consensus.Cardano.Block
+import           Ouroboros.Consensus.Cardano.CanHardFork
+import           Ouroboros.Consensus.HardFork.Combinator
+import           Ouroboros.Consensus.HardFork.Combinator.State.Types
+import           Ouroboros.Consensus.HardFork.Combinator.Util.Functors
+import           Ouroboros.Consensus.HardFork.Combinator.Util.Telescope
+import           Ouroboros.Consensus.Ledger.Tables
+import           Ouroboros.Consensus.Protocol.Praos.Translate ()
+import           Ouroboros.Consensus.Shelley.HFEras ()
+import           Ouroboros.Consensus.Shelley.Ledger.SupportsProtocol ()
+import           Ouroboros.Consensus.Shelley.ShelleyHFC (ShelleyTxOut (..))
+import           Ouroboros.Consensus.Util.SOP
+
+import qualified Cardano.Ledger.Shelley.API as SL
+
+import           Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+import           Test.Cardano.Ledger.Babbage.Serialisation.Generators ()
+
+import           Test.LedgerTables
+
+import           Test.Consensus.Byron.Generators ()
+import           Test.Consensus.Cardano.MockCrypto (MockCryptoCompatByron)
+import           Test.Consensus.Shelley.Generators ()
+
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+type Crypto = MockCryptoCompatByron
+
+tests :: TestTree
+tests = testGroup "LedgerTables"
+  [ testGroup "Cardano"
+    [ testProperty "Stowable laws" (prop_stowable_laws @(CardanoBlock Crypto))
+    , testProperty "TableStuff laws"
+        (\(InSameEra ls tbs) -> prop_tablestuff_laws @(CardanoBlock Crypto) ls tbs)
+    ]
+  ]
+
+instance Arbitrary (LedgerTables (LedgerState (CardanoBlock Crypto)) ValuesMK) where
+  arbitrary = projectLedgerTables . unstowLedgerTables <$> arbitrary
```

Comment:
We're defining the `Arbitrary` instance for `LedgerTables CardanoBlock` in terms of the larger `LedgerState CardanoBlock`. I would expect we have the arbitrary definition here and use it in the `Arbitrary` instance for `LedgerState`.

Reply:
This is not exactly the way it is. We are defining the `Arbitrary` instance for `LedgerTables CardanoBlock` in terms of the instance `Arbitrary (LedgerState (HardForkBlock xs) mk)` which defer to each generator for each era. As the Cardano tables are defined for the Cardano Block and not compositionally I don't think we can flip the relation here